### ISSUE TITLE
Just-in-Time registration of device certificates always fails

### ISSUE THREAD

**Issue by shekharhimanshu** [aws/aws-iot-device-sdk-cpp#10]
Hello,

My iot devices do [Just-in-Time registration](https://aws.amazon.com/blogs/iot/just-in-time-registration-of-device-certificates-on-aws-iot/) of device certificates when the first pub/sub mqtt command is issued. Having tried many times, I am unable to get it working with aws-iot-device-sdk-cpp. The SDK will always return `NETWORK_SSL_CONNECT_ERROR`.

Note that if I do Just-in-Time registration from some other MQTT client like, [Mosquitto](https://mosquitto.org) before using aws-iot-device-sdk-cpp, the pub/sub request succeeds.

Have you guys tested Just-in-Time registration with this SDK?
(I am using the default OpenSSLConnection provided with the SDK for my tests.)

**Comment by chaurah** [comment]
Hi @shekharhimanshu,
      That is a known issue that we discovered recently. We will be fixing the issue in an upcoming release of the SDK. We haven't gotten around to debugging it yet, but from my understanding of how JIT registration works, we need to pass in both the root CA and the device certificate during the TLS handshake for IoT to perform the registration. Currently, the SDK only passes in the device cert during the handshake. I will look into a possible fix and update this issue when I have something to unblock you.

If you want to try out a fix on your end, you can make changes to add the Root CA similar to [this ](https://github.com/aws/aws-iot-device-sdk-cpp/blob/master/network/OpenSSL/OpenSSLConnection.cpp#L262) line which adds the device cert.

Please let me know if you have any further questions.

Rahul

**Comment by shekharhimanshu** [comment]
Hello @chaurah,

Thank you for the confirmation. Yes, I have the same understanding. Not sending root CA might be the cause here.

For anyone stumbling upon the same issue, as a workaround, you can use the `MbedTLSConnection` instead of OpenSSL one. The mbedTLS implementation seems to be working well with Just-In-Time registration for now.

 I guess, sending root CA should work for `OpenSSLConnection` as well as suggested by @chaurah but I have not tested it yet. (I might test this later and update this thread)

By the way, any ETA on the next release?

**Comment by chaurah** [comment]
Hi @shekharhimanshu,
    Unfortunately I can't give out exact dates for the upcoming SDK releases. I can tell you that a fix for this particular issue will not be included in the next release. I have added it to the list of issues to be fixed for v1.2 of the SDK which should follow fairly soon after v1.1.

Rahul

**Comment by vareddy** [comment]
Hi @shekharhimanshu,
The fix for JITR is in the v1.1.1 release of the SDK. Please open another issue  if you encounter any problems.
Thanks!
Varun
